### PR TITLE
Create deferred intent using elements/session

### PR DESCRIPTION
**Pull request by porter-stripe**
## Summary
- Adds a function which can create a deferred intent on the elements/session endpoint
- Adds unit tests

## Motivation
Decoupling

## Testing
- New unit tests

## Changelog
N/A

### CODE REVIEW

File: Stripe/StripeiOSTests/STPIntentWithPreferencesTest.swift [hunk 1]
```diff
@@ -92,4 +92,39 @@ class STPIntentWithPreferencesTest: XCTestCase {
...
+                )
+
+                XCTAssertEqual(
+                    deferredIntent.paymentMethodTypes,
+                    STPPaymentMethod.types(from: ["card", "cashapp"])
+                )
+
+                XCTAssertEqual(
+                    deferredIntent.unactivatedPaymentMethodTypes,
+                    [STPPaymentMethodType.cashApp]
```
Comment (porter-stripe) [inline Stripe/StripeiOSTests/STPIntentWithPreferencesTest.swift:120]: Someone could activate cash app on our CI and cause this to fail 🤷

File: StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/STPAPIClient+PaymentSheet.swift [hunk 2]
```diff
@@ -91,6 +91,7 @@ extension STPAPIClient {
 extension STPAPIClient {
     typealias STPPaymentIntentWithPreferencesCompletionBlock = ((Result<STPPaymentIntent, Error>) -> Void)
     typealias STPSetupIntentWithPreferencesCompletionBlock = ((Result<STPSetupIntent, Error>) -> Void)
+    typealias STPElementsSessionWithPreferencesCompletionBlock = ((Result<STPElementsSession, Error>) -> Void)
```
Comment (yuki-stripe) [inline StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/STPAPIClient+PaymentSheet.swift:94]: I think we should remove "withPreferences", so `STPElementsSessionCompletionBlock`.

It doesn't make sense in the above completion blocks either, but we can leave that alone for now.

File: StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/STPAPIClient+PaymentSheet.swift [hunk 3]
```diff
@@ -108,9 +109,41 @@ extension STPAPIClient {
...
+            }
+
+            completion(.success(paymentIntentWithPreferences))
+        }
+    }
+
+    // TODO(porter) Pass in deferred intent config from public API
+    func retrieveDeferredIntentWithPreferences(
+        completion: @escaping STPElementsSessionWithPreferencesCompletionBlock
+    ) {
```
Comment (yuki-stripe) [inline StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/STPAPIClient+PaymentSheet.swift:129]: I think this should be called `retrieveElementsSession`

File: StripePayments/StripePayments/Source/API Bindings/Models/STPElementsSession.swift [hunk 4]
```diff
@@ -0,0 +1,95 @@
+//
+//  STPElementsSession.swift
+//  StripePayments
+//
+//  Created by Nick Porter on 2/15/23.
+//
+
+import Foundation
+
+public class STPElementsSession: NSObject {
```
Comment (yuki-stripe) [inline StripePayments/StripePayments/Source/API Bindings/Models/STPElementsSession.swift:10]: Let's comment this
// The response returned by v1/elements/sessions

Comment (yuki-stripe) [inline StripePayments/StripePayments/Source/API Bindings/Models/STPElementsSession.swift:10]: This needs to be `@_spi(STP)` right?  We don't want it publicly exposed

File: StripePayments/StripePayments/Source/API Bindings/Models/STPElementsSession.swift [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+
+    /// The list of payment method types (e.g. `[NSNumber(value: STPPaymentMethodType.card.rawValue)]`) that this ElementsSession is allowed to use.
+    @objc public let paymentMethodTypes: [NSNumber]
```
Comment (yuki-stripe) [inline StripePayments/StripePayments/Source/API Bindings/Models/STPElementsSession.swift:13]: `@_spi(STP)`?

File: StripePayments/StripePayments/Source/API Bindings/Models/STPElementsSession.swift [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+    /// Link-specific settings for this ElementsSession.
+    @_spi(STP) public let linkSettings: LinkSettings?
+
+    /// Country code of the user.
+    @_spi(STP) public let countryCode: String?
+
+    /// A map describing payment method types form specs.
+    @_spi(STP) public let paymentMethodSpecs: [[AnyHashable: Any]]?
+
+    @objc public let allResponseFields: [AnyHashable: Any]
```
Comment (yuki-stripe) [inline StripePayments/StripePayments/Source/API Bindings/Models/STPElementsSession.swift:30]: spi?

**Review by yuki-stripe — APPROVED** [review]

File: Stripe/StripeiOSTests/STPIntentWithPreferencesTest.swift [hunk 7]
```diff
@@ -92,4 +92,39 @@ class STPIntentWithPreferencesTest: XCTestCase {
         }
         wait(for: [expectation], timeout: STPTestingNetworkRequestTimeout)
     }
+
+    func testDeferredIntentWithPreferences() {
```
Comment (yuki-stripe) [inline Stripe/StripeiOSTests/STPIntentWithPreferencesTest.swift:96]: testRetrieveElementSession

File: Stripe/StripeiOSTests/STPIntentWithPreferencesTest.swift [hunk 8]
```diff
@@ -92,4 +92,39 @@ class STPIntentWithPreferencesTest: XCTestCase {
...
+        let expectation = XCTestExpectation(description: "Retrieve Deferred Intent With Preferences")
```
Comment (yuki-stripe) [inline Stripe/StripeiOSTests/STPIntentWithPreferencesTest.swift:97]:
```suggestion
        let expectation = XCTestExpectation(description: "Retrieve ElementsSession")
```

File: Stripe/StripeiOSTests/STPIntentWithPreferencesTest.swift (hunk 1 above)
Comment (yuki-stripe) [reply Stripe/StripeiOSTests/STPIntentWithPreferencesTest.swift:120]: Hmmm yeah I think it's okay to not test this one specifically.

File: Stripe/StripeiOSTests/STPIntentWithPreferencesTest.swift [hunk 9]
```diff
@@ -92,4 +92,39 @@ class STPIntentWithPreferencesTest: XCTestCase {
...
+                XCTAssertNotNil(deferredIntent.paymentMethodSpecs)
+
+                XCTAssertEqual(
+                    deferredIntent.orderedPaymentMethodTypes,
+                    [STPPaymentMethodType.card, STPPaymentMethodType.cashApp]
+                )
+
+                XCTAssertEqual(
+                    deferredIntent.paymentMethodTypes,
+                    STPPaymentMethod.types(from: ["card", "cashapp"])
```
Comment (yuki-stripe) [inline Stripe/StripeiOSTests/STPIntentWithPreferencesTest.swift:115]: Do we care what exactly is inside these arrays?  I think we just care that it was decoded properly, maybe we can assert that they're non-empty.

**Review by yuki-stripe — CHANGES_REQUESTED** [review]

File: StripePayments/StripePayments/Source/API Bindings/Models/STPElementsSession.swift (hunk 5 above)
Comment (porter-stripe) [reply StripePayments/StripePayments/Source/API Bindings/Models/STPElementsSession.swift:13]: Yeah, although now that the entire class is under SPI I'm not sure we need the property level ones.

Comment (yuki-stripe) [reply StripePayments/StripePayments/Source/API Bindings/Models/STPElementsSession.swift:13]: I'd guess you don't need property level ones anymore.

File: Stripe/StripeiOSTests/STPIntentWithPreferencesTest.swift (hunk 9 above)
Comment (porter-stripe) [reply Stripe/StripeiOSTests/STPIntentWithPreferencesTest.swift]: Yeah checking if empty is probably enough.

**Review by yuki-stripe — APPROVED** [review]

**Pull request merged**